deepspeed/runtime/zero/stage3.py

```diff
@@ -1260,9 +1266,10 @@ def __partition_grads(self,
         offload_fp32_gradients = {}
         offload_fp32_offsets = {}
         for param, grad_partition in zip(params_to_release, grad_partitions):
-            if param.partition_numel() * dist.get_rank(
-                    self.dp_process_group) > param.ds_numel:
+            if param.partition_numel() * (
+                    1 + dist.get_rank(self.dp_process_group)) > param.ds_numel:
```

Comment:
@stas00, turns out this fix is buggy because it will skip when the rank contains a tensor fragment and padding.

Reply:
but we added it specifically to solve the problem of another one off.

remember we had 2+6, instead of 1+7, because rank0 always returns 0, and so rank1 gets assigned the role of rank 0

where did you discover the bug - was it the existing test?

Reply:
This is the failing [test](https://github.com/microsoft/DeepSpeed/blob/81b4d5db06b70ed7bc188568f90d89af32f876b1/tests/unit/runtime/zero/test_zero.py#L532). I think we created a new [unit test](https://github.com/microsoft/DeepSpeed/blob/1b6003bd45f8fdea04e7f7729cdfc878887e70c4/tests/unit/runtime/zero/test_zero_tensor_fragment.py#L83) for the original problem, makes pass with update in my local box.

Reply:
I see. 

I still think that logic is problematic - since it tries to count ranks, but multiplies by 0 for the first rank which will always lead to a `False`  as `0 < !0` which is wrong, since the first rank's numel isn't 0.

I think it just happens to work. And we found this totally accidentally when I was debugging the hanging problem.

Reply:
So when you undid the fix we applied you actually created a one-off problem, but it's very possible the rest of the code was written to rely on the side effect of this situation and thus it was compensated on elsewhere.

but I'm not insisting that it has to be fixed. If the tests pass it's fine with me. It'll probably re-appear in the future.

Reply:
Actually, I would like to revisit the hanging problem and the overall correctness of this logic. My new sense is that the condition for skipping is when the rank only contains paddings. In that case, I think the check is whether the first byte of the rank is less than true payload. I would like to sync to chat more on this.

Reply:
the new fix we figured out is perfect, can resolve this one.